hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/util/CopyOnWriteCollection.java

```diff
@@ -0,0 +1,215 @@
+/*
+ * Copyright (c) 2008-2019, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.jet.impl.util;
+
+import javax.annotation.Nonnull;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;
+
+/**
+ * Non-blocking thread-safe collection returning a refreshable iterator. The collection is biased towards fast repeated
+ * iteration over all stored elements.
+ *
+ * Some of the mutation methods are currently not implemented - they can be added as needed.
+ *
+ * @param <T>
+ */
+public final class CopyOnWriteCollection<T> implements Collection<T> {
```

Comment:
Was there a need to reimplement `CopyOnWriteArrayList`? We could build the `RefreshableIterator` on top of `CopyOnWriteArrayList.getArray()`.

Reply:
True, i noticed it's not private, but missed it's also not public :)